docs/tutorials/pytorch.ipynb

```diff
@@ -0,0 +1,507 @@
+    "* ML: when machines learn from example\n",
+    "* DL: when machines learn using neural networks\n",
+    "\n",
+    "In this definition, DL is a subset of ML, and ML is a subset of AI. Some common examples of models and applications of these include:\n",
+    "\n",
+    "* AI: Minimax, A*, Deep Blue, video game AI\n",
+    "* ML: OLS, SVM, $k$-means, spam filtering\n",
+    "* DL: MLP, CNN, ChatGPT, self-driving cars\n",
+    "\n",
+    "In this tutorial, we will specifically focus on deep learning, but many of the same concepts are shared with machine learning."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7f26e4b8",
+   "metadata": {
+    "id": "5rLknZxrBEMz"
+   },
+   "source": [
+    "## Datasets\n",
+    "\n",
+    "In order to learn by example, we first need examples. In machine learning, we construct datasets of the form:\n",
+    "\n",
+    "$$\\mathcal{D} = \\left\\{\\left(x^{(i)}, y^{(i)}\\right)\\right\\}_{i=1}^N$$\n",
+    "\n",
+    "Written in English, dataset $\\mathcal{D}$ is composed of $N$ pairs of inputs $x$ and expected outputs $y$. $x$ and $y$ can be tabular data, images, text, or any other object that can be represented mathematically.\n",
+    "\n",
+    "![EuroSAT](https://github.com/phelber/EuroSAT/blob/master/eurosat-overview.png?raw=true)\n",
+    "\n",
+    "In this tutorial (and many later tutorials), we will use EuroSAT100, a toy dataset composed of 100 images from the [EuroSAT](https://github.com/phelber/EuroSAT) dataset. EuroSAT is a popular image classification dataset with multispectral images from the Sentinel-2 satellites. Each image is classified into one of ten categories or \"classes\":\n",
+    "\n",
+    "0. Annual Crop\n",
+    "1. Forest\n",
+    "2. Herbaceous Vegetation\n",
+    "3. Highway\n",
+    "4. Industrial Buildings\n",
+    "5. Pasture\n",
+    "6. Permanent Crop\n",
+    "7. Residential Buildings\n",
+    "8. River\n",
+    "9. Sea & Lake\n",
+    "\n",
+    "We can load this dataset and visualize the RGB bands of some example $(x, y)$ pairs like so:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "fa0c5a0c-ac4c-44c5-9fb7-fe4be07a0f01",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "root = os.path.join(tempfile.gettempdir(), 'eurosat100')\n",
+    "dataset = EuroSAT100(root, download=True)\n",
+    "\n",
+    "for i in torch.randint(len(dataset), (10,)):\n",
+    "    sample = dataset[i]\n",
+    "    dataset.plot(sample)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f89e20ae-d3b6-4f05-a83f-f7034dd9862f",
+   "metadata": {},
+   "source": [
+    "In machine learning, we not only want to train a model, but also evaluate its performance on unseen data. Oftentimes, our dataset is split into three separate subsets:\n",
+    "\n",
+    "* train: for training the model *parameters*\n",
+    "* val: for validating the model *hyperparameters*\n",
+    "* test: for testing the model *performance*\n",
```

Comment:
Validation is also to assess model performance right? Strictly speaking, testing data should only be run once at the very end to "deploy" the model, although almost everyone is indirectly tuning on test set performance as well.

Reply:
Technically, you're only supposed to use validation for hyperparameter tuning (and early stopping). You do evaluate performance on val, but only to decide which set of hyperparams is best. I think I've used it as such in the rest of the tutorial.

It may be worth noting that sometimes test isn't public, but I didn't go into that much detail here.